kortex_driver/src/non-generated/driver/kortex_arm_simulation.cpp

```diff
@@ -514,76 +645,402 @@ void KortexArmSimulation::PlayAction(const kortex_driver::Action& action)
+    // Get Start frame
+    // For the Rotation part : ThetaX = gamma, ThetaY = beta, ThetaZ = alpha 
+    auto start = KDL::Frame();
+    Eigen::VectorXd positions_eigen(m_degrees_of_freedom);
+    for (int i = 0; i < GetDOF(); i++)
+    {
+        positions_eigen[i] = current.position[m_first_arm_joint_index + i];
+    }
+    KDL::JntArray current_kdl(GetDOF());
+    current_kdl.data = positions_eigen;
+    m_fk_solver->JntToCart(current_kdl, start);
+
+    {
+    ROS_DEBUG("START FRAME :");
+    ROS_DEBUG("X=%2.4f Y=%2.4f Z=%2.4f", start.p[0], start.p[1], start.p[2]);
+    double sa, sb, sg; start.M.GetEulerZYX(sa, sb, sg);
+    ROS_DEBUG("ALPHA=%2.4f BETA=%2.4f GAMMA=%2.4f", m_math_util.toDeg(sa), m_math_util.toDeg(sb), m_math_util.toDeg(sg));
+    KDL::Vector axis;
+    ROS_DEBUG("start rot = %2.4f", start.M.GetRotAngle(axis));
+    }
+
+    // Get End frame
+    auto end_pos = KDL::Vector(constrained_pose.target_pose.x, constrained_pose.target_pose.y, constrained_pose.target_pose.z);
+    auto end_rot = KDL::Rotation::EulerZYX(m_math_util.toRad(constrained_pose.target_pose.theta_z), m_math_util.toRad(constrained_pose.target_pose.theta_y), m_math_util.toRad(constrained_pose.target_pose.theta_x));
+    KDL::Frame end(end_rot, end_pos);
+
+    {
+    ROS_DEBUG("END FRAME :");
+    ROS_DEBUG("X=%2.4f Y=%2.4f Z=%2.4f", end_pos[0], end_pos[1], end_pos[2]);
+    double ea, eb, eg; end_rot.GetEulerZYX(ea, eb, eg);
+    ROS_DEBUG("ALPHA=%2.4f BETA=%2.4f GAMMA=%2.4f", m_math_util.toDeg(ea), m_math_util.toDeg(eb), m_math_util.toDeg(eg));
+    KDL::Vector axis;
+    ROS_DEBUG("end rot = %2.4f", end_rot.GetRotAngle(axis));
+    }
+
+    // If different speed limits than the default ones are provided, use them instead
+    float translation_speed_limit = m_max_cartesian_twist_linear;
+    float rotation_speed_limit = m_max_cartesian_twist_angular;
+    if (!constrained_pose.constraint.oneof_type.speed.empty())
+    {
+        // If a max velocity is supplied for each joint, we need to find the limiting duration with this velocity constraint
+        translation_speed_limit = std::min(constrained_pose.constraint.oneof_type.speed[0].translation, translation_speed_limit);
+        rotation_speed_limit = std::min(constrained_pose.constraint.oneof_type.speed[0].orientation, rotation_speed_limit);
+    }
+
+    // Calculate norm of translation movement and minimum duration to move this amount given max translation speed
+    double delta_pos = (end_pos - start.p).Norm();
+    double minimum_translation_duration = delta_pos / m_max_cartesian_twist_linear; // in seconds
+
+    // Calculate angle variation of rotation movement and minimum duration to move this amount given max rotation speed
+    KDL::Vector axis; // we need to create this variable to access the RotAngle for start and end frames'rotation components
+    double delta_rot = fabs(end_rot.GetRotAngle(axis) - start.M.GetRotAngle(axis));
```

Comment:
The correct way to compute rotation angle angle between two rotation matrices would be to: 
1 - Compute rotation leading from start.M to end_rot -> end_rot = dR * start.M -> dR = end_rot*start.M.transpose()
2 - Compute delta_rot -> double delta_rot = dR.GetRotAngle(axis);

Reply:
Done